File: src/main/kotlin/dev/blackoutburst/game/texture/TextureArray.kt
package dev.blackoutburst.game.texture

import dev.blackoutburst.game.utils.IOUtils
import dev.blackoutburst.game.utils.stack
import org.lwjgl.opengl.GL30.*
import org.lwjgl.stb.STBImage
import java.nio.ByteBuffer

class TextureArray(files: List<String>, size: Int = 16) {
    var id = -1

    init {
        stack { stack ->
            val widthBuffer = stack.mallocInt(1)
            val heightBuffer = stack.mallocInt(1)
            val comp = stack.mallocInt(1)

            id = glGenTextures()
            glBindTexture(GL_TEXTURE_2D_ARRAY, id)
            glTexImage3D(
                GL_TEXTURE_2D_ARRAY, 0, GL_RGBA8,
                size, size, files.size, 0,
                GL_RGBA, GL_UNSIGNED_BYTE, null as ByteBuffer?
            )

            for ((index, file) in files.withIndex()) {
                val data = STBImage.stbi_load_from_memory(IOUtils.ioResourceToByteBuffer(file, 1024), widthBuffer, heightBuffer, comp, 4)
                    ?: continue

                glTexSubImage3D(
                    GL_TEXTURE_2D_ARRAY, 0, 0, 0, index,
                    size, size, 1, GL_RGBA, GL_UNSIGNED_BYTE, data
                )
                glGenerateMipmap(GL_TEXTURE_2D_ARRAY)
                STBImage.stbi_image_free(data)
            }

            glTexParameteri(GL_TEXTURE_2D_ARRAY, GL_TEXTURE_MIN_FILTER, GL_NEAREST_MIPMAP_NEAREST)
            glTexParameteri(GL_TEXTURE_2D_ARRAY, GL_TEXTURE_MAG_FILTER, GL_NEAREST)
            glTexParameteri(GL_TEXTURE_2D_ARRAY, GL_TEXTURE_WRAP_S, GL_REPEAT)
            glTexParameteri(GL_TEXTURE_2D_ARRAY, GL_TEXTURE_WRAP_T, GL_REPEAT)
        }
    }
}

File: C/src/graphics/textureArray.c
#define GLFW_INCLUDE_NONE

#include <stdio.h>
#include <stdlib.h>
#include "utils/types.h"
#include "glfw/glfw3.h"
#include "graphics/opengl.h"
#include "utils/math.h"
#include "stb/stb_image.h"

void textureArrayDelete(U32 id) {
    glDeleteTextures(1, &id);
}

U32 textureArrayCreate(I8** files, I16 filesCount, I16 fileSize) {
    U32 id = 0;
    glGenTextures(1, &id);

    glBindTexture(GL_TEXTURE_2D_ARRAY, id);
    glTexImage3D(GL_TEXTURE_2D_ARRAY, 0, GL_RGBA8, fileSize, fileSize, filesCount, 0, GL_RGBA, GL_UNSIGNED_BYTE, NULL);

    I32 w = 0;
    I32 h = 0;
    I32 comp = 0;
    for (I32 i = 0; i < filesCount; i++) {
        U8* data = stbi_load(files[i], &w, &h, &comp, 4);
        if (!data) {
            printf("Failed to load image %s\n", files[i]);
            return 0;
        }

        glTexSubImage3D(GL_TEXTURE_2D_ARRAY, 0, 0, 0, i, fileSize, fileSize, 1, GL_RGBA, GL_UNSIGNED_BYTE, data);
        glGenerateMipmap(GL_TEXTURE_2D_ARRAY);
        stbi_image_free(data);
    }

    glTexParameteri(GL_TEXTURE_2D_ARRAY, GL_TEXTURE_MIN_FILTER, GL_NEAREST_MIPMAP_NEAREST);
    glTexParameteri(GL_TEXTURE_2D_ARRAY, GL_TEXTURE_MAG_FILTER, GL_NEAREST);
    glTexParameteri(GL_TEXTURE_2D_ARRAY, GL_TEXTURE_WRAP_S, GL_REPEAT);
    glTexParameteri(GL_TEXTURE_2D_ARRAY, GL_TEXTURE_WRAP_T, GL_REPEAT);

    return id;
}
